llvm/tools/llvm-config/CMakeLists.txt

```diff
@@ -64,7 +64,11 @@ set(LLVM_CFLAGS "${LLVM_C_STD_FLAG} ${LLVM_DEFINITIONS}")
 set(LLVM_CXXFLAGS "${CMAKE_CXX${CMAKE_CXX_STANDARD}_STANDARD_COMPILE_OPTION} ${LLVM_CXX_STDLIB_FLAG} ${COMPILE_FLAGS} ${LLVM_DEFINITIONS}")
 set(LLVM_BUILD_SYSTEM cmake)
 set(LLVM_HAS_RTTI ${LLVM_CONFIG_HAS_RTTI})
-set(LLVM_DYLIB_VERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}")
+if(CYGWIN)
+  set(LLVM_DYLIB_VERSION "${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}${LLVM_VERSION_SUFFIX}")
+else()
+  set(LLVM_DYLIB_VERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}")
+endif()
```

Comment:
Maybe it would be better if Cygwin just followed other targets?

Reply:
The question is which other targets?  UNIXy targets put major and minor in the filename of the shared library (libLLVM.so.20.1).  MinGW targets puts just the major (libLLVM-20.dll).  Neither appear in the import library name (which is admittedly a PE/COFF thing).

Reply:
It would be less invasive to llvm-config if the cmake files were convinced to generate cygLLVM-20.dll and libLLVM-20.dll.a.

Reply:
I meant other Win32 targets (so probably only MinGW?), but indeed that looks to be a mess.
On Linux, you've got more variants, so the flexibility is higher:
```
❯ pacman -Ql llvm-libs | rg libLLVM
llvm-libs /usr/lib/libLLVM-19.so  # <- symlink
llvm-libs /usr/lib/libLLVM.so # <- symlink
llvm-libs /usr/lib/libLLVM.so.19.1 # <- real lib
```
I think following MinGW will be less surprising, though.

Reply:
yeah I don't really get why it's not making the symlinks on Cygwin (which defaults to real (WSL-style) symlinks)

Reply:
@tyan0 what do you think about how the soversion should be handled on Cygwin?

Reply:
I was just looking at @tyan0's patches for proposed Cygwin patches and saw:
```patch
--- origsrc/llvm-20.1.2.src/tools/llvm-shlib/CMakeLists.txt	2025-04-11 08:58:02.655364700 +0900
+++ src/llvm-20.1.2.src/tools/llvm-shlib/CMakeLists.txt	2025-04-11 23:15:33.171264700 +0900
@@ -33,7 +33,7 @@ if(LLVM_BUILD_LLVM_DYLIB)
   if (LLVM_LINK_LLVM_DYLIB)
     set(INSTALL_WITH_TOOLCHAIN INSTALL_WITH_TOOLCHAIN)
   endif()
-  if (WIN32)
+  if (WIN32 OR CYGWIN)
     add_llvm_library(LLVM SHARED DISABLE_LLVM_LINK_LLVM_DYLIB SONAME ${INSTALL_WITH_TOOLCHAIN} ${SOURCES})
   else()
     add_llvm_library(LLVM SHARED DISABLE_LLVM_LINK_LLVM_DYLIB OUTPUT_NAME LLVM ${INSTALL_WITH_TOOLCHAIN} ${SOURCES})
```

This is what I am trying now to see if that makes the output look like MinGW.

Reply:
With that patch, you get
* bin/
  * cygclang-20.1.dll
  * cygclang-cpp-20.1.dll
  * cygLLVM-20.dll
  * cygRemarks-20.1.dll
* lib/
  * libclang.dll.a
  * libclang-cpp.dll.a
  * libLLVM-20.dll.a
  * libRemarks.dll.a
  * cygLLVM.dll -> cygLLVM-20.dll
  * cygLLVM-20.1.3.dll -> cygLLVM-20.dll
  * cygLLVM-20.dll -> cygLLVM-20.1.dll

The last three are broken symlinks, there is no cygLLVM-20.1.dll, and it would be in bin if there were (where the DLLs should be).  If something was going to be symlinked, probaby the import library would be the better choice, so that passing `-l` of them will work, but all resolve to the same dll.

Reply:
> I don't like the idea of having libLLVM diverge from libclang/libclang-cpp though.

That does seem to be the case everywhere right now anyway, so that's more of an upstream LLVM-in-general thing?

I think that trying to mimic the mingw behaviour here probably is the most reasonable thing to do. (The broken symlinks is something that doesn't really show up in mingw cases I think?)

Tests for changes to this tool are in `llvm/test/tools/llvm-config`. But I'm not sure if anything of this really is covered there so far, as this is quite dependent on the type of build (whether dylibs are enabled etc). See `system-libs.windows.test` and `system-libs.test` for host OS specific tests there.

Reply:
> I think that trying to mimic the mingw behaviour here probably is the most reasonable thing to do. (The broken symlinks is something that doesn't really show up in mingw cases I think?)

The broken symlinks are in an `if (UNIX)` block, which I now patched to be `if (UNIX AND NOT CYGWIN)`, and also patched the function making the symlinks to know shared libs live in `BINDIR` on `CYGWIN` as well as `WIN32`.